biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/JUnitEclipseListener.java

```diff
@@ -0,0 +1,528 @@
+			for (TestIdentifier child : testPlan.getChildren(root)) {
+				visitEntry(child);
+			}
+		}
+		startTime = System.currentTimeMillis();
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		message("%RUNTIME", "" + (System.currentTimeMillis() - startTime));
+		out.flush();
+		if (controlSock == null) {
+			setNullStreams();
+		}
+	}
+
+	private void sendTrace(Throwable t) {
+		message("%TRACES ");
+		t.printStackTrace(out);
+		if (verbose) {
+			t.printStackTrace(System.err);
+		}
+		out.println();
+		message("%TRACEE ");
+	}
+
+	private void sendExpectedAndActual(CharSequence expected, CharSequence actual) {
+		message("%EXPECTS");
+		out.println(expected);
+		info(expected);
+		message("%EXPECTE");
+
+		message("%ACTUALS");
+		out.println(actual);
+		info(actual);
+		message("%ACTUALE");
+	}
+
+	@Override
+	public void executionStarted(TestIdentifier test) {
+		info("JUnitEclipseListener: Execution started: " + test);
+		if (test.isTest()) {
+			message("%TESTS  ", test);
+		}
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier test, TestExecutionResult testExecutionResult) {
+		info("JUnitEclipseListener: Execution finished: " + test);
+		Status result = testExecutionResult.getStatus();
+		if (test.isTest()) {
+			if (result != Status.SUCCESSFUL) {
+				final boolean assumptionFailed = result == Status.ABORTED;
+				info("JUnitEclipseListener: assumption failed: " + assumptionFailed);
+				Optional<Throwable> throwableOp = testExecutionResult.getThrowable();
+				if (throwableOp.isPresent()) {
+					Throwable exception = throwableOp.get();
+					info("JUnitEclipseListener: throwable: " + exception);
+
+					if (assumptionFailed || exception instanceof AssertionError) {
+						info(() -> "JUnitEclipseListener: failed: " + exception + " assumptionFailed: "
+							+ assumptionFailed);
+						message("%FAILED ", test, (assumptionFailed ? "@AssumptionFailure: " : ""));
+
+						sendExpectedAndActual(exception);
+
+					} else {
+						info("JUnitEclipseListener: error");
+						message("%ERROR  ", test);
+					}
+					sendTrace(exception);
+				}
+			}
+			message("%TESTE  ", test);
+		} else { // container
+			if (result != Status.SUCCESSFUL) {
+				message("%ERROR  ", test);
+				Optional<Throwable> throwableOp = testExecutionResult.getThrowable();
+				if (throwableOp.isPresent()) {
+					sendTrace(throwableOp.get());
+				}
+			}
+		}
+	}
+
+	private static void appendString(StringBuilder b, String s) {
+		if (b.length() > 0) {
+			b.append("\n\n");
+		}
+		b.append(s);
+	}
+
+	private boolean sendExpectedAndActual(Throwable exception, StringBuilder expectedBuilder,
+		StringBuilder actualBuilder) {
+		BooleanSupplier b;
+		// switch is based on the class name rather than using instanceof
+		// to avoid hard dependency on the assertion types.
+		switch (exception.getClass()
+			.getName()) {
+			case "org.opentest4j.AssertionFailedError" :
+				b = () -> {
```

Comment:
Do we really need to use lambda here? The lambdas never escape the method. You could simply just put the lambda bodies as the switch blocks.

Reply:
If the code is inlined it introduces a hard dependency on the assertion classes, which means that JUnitEclipseListener won't load unless JUnit 4 is present. The lambdas prevent a hard dependency on the assertion classes so I can make the JUnit 4 import optional. I could possibly inline the OpenTest4J one though, a I can't make that optional due to dependencies elsewhere, but I did then all the same got consistency.

Reply:
"but I did them all the same for consistency." I have to stop responding on my phone... :smile:

Reply:
OK, then please add a comment as to why the code is this way. Otherwise someone may come along and "simplify" things which will add the undesired hard dependencies.

Reply:
I have added a comment.

Also, if they do attempt to simplify it or do anything else in the code re-introduces a hard dependency on JUnit 3/4, then these unit tests will fail:
https://github.com/bndtools/bnd/blob/14f8f6c9d94bfcc8d4af0f0f0fc814250c8dbf64/biz.aQute.tester.test/test/aQute/tester/junit/platform/test/ActivatorTest.java#L319
https://github.com/bndtools/bnd/blob/14f8f6c9d94bfcc8d4af0f0f0fc814250c8dbf64/biz.aQute.tester.test/test/aQute/tester/junit/platform/test/ActivatorTest.java#L344

So we shouldn't get any accidental regressions.